File: tests/test_memory_plan.py
"""Test suite for buildMemoryPlan greedy buffer assignment algorithm.
Reimplements the C++ algorithm in Python to verify correctness."""

import sys

def build_memory_plan(intervals):
    """intervals: list of dict {name, firstUse, lastUse, sizeBytes}"""
    order = sorted(range(len(intervals)), key=lambda i: intervals[i]["sizeBytes"], reverse=True)

    slots = []  # list of {id, sizeBytes, occupied: [(first, last)]}
    plan = {}

    for idx in order:
        iv = intervals[idx]
        best_slot = -1
        best_waste = float("inf")
        for s_idx, slot in enumerate(slots):
            if slot["sizeBytes"] < iv["sizeBytes"]:
                continue
            overlaps = any(iv["firstUse"] <= b and iv["lastUse"] >= a for a, b in slot["occupied"])
            if overlaps:
                continue
            waste = slot["sizeBytes"] - iv["sizeBytes"]
            if waste < best_waste:
                best_waste = waste
                best_slot = s_idx

        if best_slot >= 0:
            slots[best_slot]["occupied"].append((iv["firstUse"], iv["lastUse"]))
            plan[iv["name"]] = {"slotId": slots[best_slot]["id"], "slotSizeBytes": slots[best_slot]["sizeBytes"]}
        else:
            sid = len(slots)
            slots.append({"id": sid, "sizeBytes": iv["sizeBytes"], "occupied": [(iv["firstUse"], iv["lastUse"])]})
            plan[iv["name"]] = {"slotId": sid, "slotSizeBytes": iv["sizeBytes"]}

    return plan, slots


passed = 0
failed = 0

def test(name, condition, detail=""):
    global passed, failed
    if condition:
        passed += 1
        print(f"  PASS: {name}")
    else:
        failed += 1
        print(f"  FAIL: {name} — {detail}")


# AC1: BufferAssignment has slotId and slotSizeBytes
plan, _ = build_memory_plan([{"name": "t0", "firstUse": 0, "lastUse": 2, "sizeBytes": 1024}])
test("AC1: struct fields", "slotId" in plan["t0"] and "slotSizeBytes" in plan["t0"])

# AC2: returns map<string, BufferAssignment>
intervals = [
    {"name": "t0", "firstUse": 0, "lastUse": 2, "sizeBytes": 1024},
    {"name": "t1", "firstUse": 1, "lastUse": 3, "sizeBytes": 2048},
]
plan, _ = build_memory_plan(intervals)
test("AC2: returns map with all tensors", len(plan) == 2 and "t0" in plan and "t1" in plan)

# AC3: Non-overlapping reuse
intervals = [
    {"name": "t0", "firstUse": 0, "lastUse": 1, "sizeBytes": 1024},
    {"name": "t1", "firstUse": 2, "lastUse": 3, "sizeBytes": 1024},
]
plan, _ = build_memory_plan(intervals)
test("AC3a: non-overlapping reuse same slot", plan["t0"]["slotId"] == plan["t1"]["slotId"])

# AC3: Overlapping → separate slots
intervals = [
    {"name": "t0", "firstUse": 0, "lastUse": 2, "sizeBytes": 1024},
    {"name": "t1", "firstUse": 1, "lastUse": 3, "sizeBytes": 1024},
]
plan, _ = build_memory_plan(intervals)
test("AC3b: overlapping get separate slots", plan["t0"]["slotId"] != plan["t1"]["slotId"])

# AC3: Largest-first ordering
intervals = [
    {"name": "t_small", "firstUse": 2, "lastUse": 3, "sizeBytes": 1024},
    {"name": "t_big", "firstUse": 0, "lastUse": 1, "sizeBytes": 4096},
]
plan, _ = build_memory_plan(intervals)
test("AC3c: largest first, small reuses big slot",
     plan["t_small"]["slotId"] == plan["t_big"]["slotId"] and plan["t_small"]["slotSizeBytes"] == 4096)

# AC3: Best-fit selection
intervals = [
    {"name": "big", "firstUse": 0, "lastUse": 1, "sizeBytes": 8192},
    {"name": "medium", "firstUse": 0, "lastUse": 1, "sizeBytes": 4096},
    {"name": "small", "firstUse": 2, "lastUse": 3, "sizeBytes": 3000},
]
plan, _ = build_memory_plan(intervals)
test("AC3d: best-fit picks least waste", plan["small"]["slotId"] == plan["medium"]["slotId"])

# AC4: Transformer-like achieves ≤50% of naive
intervals = []
tensor_size = 4 * 1024 * 1024
for layer in range(12):
    base = layer * 3
    intervals.append({"name": f"attn_{layer}", "firstUse": base, "lastUse": base + 1, "sizeBytes": tensor_size})
    intervals.append({"name": f"ffn1_{layer}", "firstUse": base + 1, "lastUse": base + 2, "sizeBytes": tensor_size})
    intervals.append({"name": f"ffn2_{layer}", "firstUse": base + 2, "lastUse": base + 3, "sizeBytes": tensor_size})

plan, slots = build_memory_plan(intervals)
total_naive = sum(iv["sizeBytes"] for iv in intervals)
total_planned = sum(s["sizeBytes"] for s in slots)
ratio = total_planned / total_naive
test(f"AC4: transformer ≤50% (got {ratio*100:.1f}%)", ratio <= 0.50)

# Edge cases
plan, _ = build_memory_plan([])
test("Edge: empty input", len(plan) == 0)

plan, _ = build_memory_plan([{"name": "only", "firstUse": 0, "lastUse": 5, "sizeBytes": 2048}])
test("Edge: single tensor", len(plan) == 1 and plan["only"]["slotId"] == 0 and plan["only"]["slotSizeBytes"] == 2048)

print(f"\n{passed}/{passed+failed} tests passed.")
sys.exit(0 if failed == 0 else 1)

File: tests/test_memory_plan.cpp
#include <algorithm>
#include <cassert>
#include <cstdio>
#include <map>
#include <numeric>
#include <string>
#include <vector>

// Copy structs from execution_context.h to make this self-contained
struct LifetimeInterval {
    std::string name;
    int firstUse;
    int lastUse;
    size_t sizeBytes;
};

struct BufferAssignment {
    int slotId;
    size_t slotSizeBytes;
};

// Copy of GraphExecutor::buildMemoryPlan
std::map<std::string, BufferAssignment> buildMemoryPlan(
    const std::vector<LifetimeInterval>& intervals) {

    std::vector<size_t> order(intervals.size());
    std::iota(order.begin(), order.end(), 0);
    std::sort(order.begin(), order.end(), [&](size_t a, size_t b) {
        return intervals[a].sizeBytes > intervals[b].sizeBytes;
    });

    struct Slot {
        int id;
        size_t sizeBytes;
        std::vector<std::pair<int, int>> occupied;
    };
    std::vector<Slot> slots;

    std::map<std::string, BufferAssignment> plan;
    size_t totalIndividual = 0;

    for (size_t idx : order) {
        auto& iv = intervals[idx];
        totalIndividual += iv.sizeBytes;

        int bestSlot = -1;
        size_t bestWaste = SIZE_MAX;
        for (size_t s = 0; s < slots.size(); ++s) {
            if (slots[s].sizeBytes < iv.sizeBytes) continue;
            bool overlaps = false;
            for (auto& [a, b] : slots[s].occupied) {
                if (iv.firstUse <= b && iv.lastUse >= a) {
                    overlaps = true;
                    break;
                }
            }
            if (overlaps) continue;
            size_t waste = slots[s].sizeBytes - iv.sizeBytes;
            if (waste < bestWaste) {
                bestWaste = waste;
                bestSlot = static_cast<int>(s);
            }
        }

        if (bestSlot >= 0) {
            slots[bestSlot].occupied.push_back({iv.firstUse, iv.lastUse});
            plan[iv.name] = {slots[bestSlot].id, slots[bestSlot].sizeBytes};
        } else {
            int id = static_cast<int>(slots.size());
            slots.push_back({id, iv.sizeBytes, {{iv.firstUse, iv.lastUse}}});
            plan[iv.name] = {id, iv.sizeBytes};
        }
    }

    return plan;
}

int tests_run = 0;
int tests_passed = 0;

#define TEST(name) \
    do { \
        tests_run++; \
        fprintf(stderr, "  TEST: %s ... ", #name); \
    } while(0)

#define PASS() do { tests_passed++; fprintf(stderr, "PASSED\n"); } while(0)
#define FAIL(msg) do { fprintf(stderr, "FAILED: %s\n", msg); } while(0)

// AC1: BufferAssignment struct has slotId and slotSizeBytes
void test_struct_fields() {
    TEST(struct_fields);
    BufferAssignment ba{3, 1024};
    assert(ba.slotId == 3);
    assert(ba.slotSizeBytes == 1024);
    PASS();
}

// AC2: buildMemoryPlan returns map<string, BufferAssignment>
void test_returns_map() {
    TEST(returns_map);
    std::vector<LifetimeInterval> intervals = {
        {"t0", 0, 2, 1024},
        {"t1", 1, 3, 2048},
    };
    auto plan = buildMemoryPlan(intervals);
    assert(plan.size() == 2);
    assert(plan.count("t0") == 1);
    assert(plan.count("t1") == 1);
    PASS();
}

// AC3: Greedy algorithm — non-overlapping tensors reuse slots
void test_non_overlapping_reuse() {
    TEST(non_overlapping_reuse);
    // t0: [0,1] 1024B, t1: [2,3] 1024B — should share a slot
    std::vector<LifetimeInterval> intervals = {
        {"t0", 0, 1, 1024},
        {"t1", 2, 3, 1024},
    };
    auto plan = buildMemoryPlan(intervals);
    assert(plan["t0"].slotId == plan["t1"].slotId);
    PASS();
}

// AC3: Overlapping tensors get different slots
void test_overlapping_separate() {
    TEST(overlapping_separate);
    std::vector<LifetimeInterval> intervals = {
        {"t0", 0, 2, 1024},
        {"t1", 1, 3, 1024},
    };
    auto plan = buildMemoryPlan(intervals);
    assert(plan["t0"].slotId != plan["t1"].slotId);
    PASS();
}

// AC3: Largest-first ordering — large tensor gets its own slot, smaller fits in existing
void test_largest_first() {
    TEST(largest_first);
    // t_big: [0,1] 4096B, t_small: [2,3] 1024B
    // t_big processed first (larger), t_small can reuse its slot
    std::vector<LifetimeInterval> intervals = {
        {"t_small", 2, 3, 1024},
        {"t_big", 0, 1, 4096},
    };
    auto plan = buildMemoryPlan(intervals);
    assert(plan["t_small"].slotId == plan["t_big"].slotId);
    assert(plan["t_small"].slotSizeBytes == 4096); // uses big slot
    PASS();
}

// AC4: Transformer-like pattern achieves <50% of naive
void test_transformer_memory_savings() {
    TEST(transformer_memory_savings);
    // Simulate a transformer: many intermediate tensors with short lifetimes
    // 12 layers, each producing 3 intermediates that die before next layer
    std::vector<LifetimeInterval> intervals;
    size_t tensorSize = 4 * 1024 * 1024; // 4MB each
    for (int layer = 0; layer < 12; layer++) {
        int base = layer * 3;
        intervals.push_back({"attn_" + std::to_string(layer), base, base + 1, tensorSize});
        intervals.push_back({"ffn1_" + std::to_string(layer), base + 1, base + 2, tensorSize});
        intervals.push_back({"ffn2_" + std::to_string(layer), base + 2, base + 3, tensorSize});
    }
    // 36 tensors × 4MB = 144MB naive
    auto plan = buildMemoryPlan(intervals);

    size_t totalNaive = 0;
    for (auto& iv : intervals) totalNaive += iv.sizeBytes;

    // Count unique slots and their sizes
    std::map<int, size_t> slotSizes;
    for (auto& [name, ba] : plan) {
        slotSizes[ba.slotId] = ba.slotSizeBytes;
    }
    size_t totalPlanned = 0;
    for (auto& [id, sz] : slotSizes) totalPlanned += sz;

    double ratio = (double)totalPlanned / totalNaive;
    fprintf(stderr, "(ratio=%.1f%%) ", ratio * 100);
    assert(ratio <= 0.50);
    PASS();
}

// Edge case: empty input
void test_empty_input() {
    TEST(empty_input);
    auto plan = buildMemoryPlan({});
    assert(plan.empty());
    PASS();
}

// Edge case: single tensor
void test_single_tensor() {
    TEST(single_tensor);
    std::vector<LifetimeInterval> intervals = {{"only", 0, 5, 2048}};
    auto plan = buildMemoryPlan(intervals);
    assert(plan.size() == 1);
    assert(plan["only"].slotId == 0);
    assert(plan["only"].slotSizeBytes == 2048);
    PASS();
}

// Best-fit: picks slot with least waste
void test_best_fit() {
    TEST(best_fit);
    // Create two slots of different sizes, then a small tensor that fits both
    std::vector<LifetimeInterval> intervals = {
        {"big", 0, 1, 8192},
        {"medium", 0, 1, 4096},
        {"small", 2, 3, 3000}, // fits in both slots, should pick medium (less waste)
    };
    auto plan = buildMemoryPlan(intervals);
    assert(plan["small"].slotId == plan["medium"].slotId);
    PASS();
}

int main() {
    fprintf(stderr, "Running buildMemoryPlan tests:\n");
    test_struct_fields();
    test_returns_map();
    test_non_overlapping_reuse();
    test_overlapping_separate();
    test_largest_first();
    test_transformer_memory_savings();
    test_empty_input();
    test_single_tensor();
    test_best_fit();
    fprintf(stderr, "\n%d/%d tests passed.\n", tests_passed, tests_run);
    return tests_passed == tests_run ? 0 : 1;
}
